adds node_id to Event::Payment{Received, Claimed}

tee8z opened the pull request:
- adds 'our_node_id' to 'Event::Payment{Received,Claimed}' and includes updates to functional test cases

Hopefully we were looking for 'our_node_id' and not the counterparty, but please correct me if that assumption is wrong. Also, right now the node_id is wrapped in an Option just to make the default easier, but can look into making it so it is not wrapped. Thank you for any review.

tee8z commented:
Both very valid points, updated code based on feedback

tnull reviewed (commented):
Generally looks good.

> Also, right now the node_id is wrapped in an Option just to make the default easier, but can look into making it so it is not wrapped. 

Are you referring to a default value for serialization? If so, you should be able to make the field a `PublicKey` and use the `Option<PublicKey>` wrapping only in a temporary variable in the serialization logic.

tee8z commented:
> Generally looks good.
> 
> > Also, right now the node_id is wrapped in an Option just to make the default easier, but can look into making it so it is not wrapped.
> 
> Are you referring to a default value for serialization? If so, you should be able to make the field a `PublicKey` and use the `Option<PublicKey>` wrapping only in a temporary variable in the serialization logic.

I changed the logic to follow this, it seems I either need to make the 'our_node_Id' be an Option<> and mark it as 'option' in write_tlv_fields! & read_tlv_fields! or have it not be wrapped and set it as 'required'. Please let me know what the preferred approach is here, I can see a case for either.

tnull commented:
> I changed the logic to follow this, it seems I either need to make the 'our_node_Id' be an Option<> and mark it as 'option' in write_tlv_fields! & read_tlv_fields! or have it not be wrapped and set it as 'required'. Please let me know what the preferred approach is here, I can see a case for either.

Ugh, I think you're right. We're probably just should to make it an `Option`, even if it's kind of unfortunate for a field that will always be there going forward. Sorry for the confusion.

tee8z commented:
Reverted the last commit to keep the 'our_node_id' as optional  to not break existing clients of the library.

tnull commented on lightning/src/util/events.rs:
```diff
@@ -249,6 +249,8 @@ pub enum Event {
 	/// [`ChannelManager::claim_funds`]: crate::ln::channelmanager::ChannelManager::claim_funds
 	/// [`ChannelManager::fail_htlc_backwards`]: crate::ln::channelmanager::ChannelManager::fail_htlc_backwards
 	PaymentReceived {
+		/// The node that recieved the payment, ie our node
```
nit:
```suggestion
		/// The ID of the node that received the payment, i.e., our node.
```

jurvis replied:
Is this field always going to be populated with "our" node id? 

I was wondering if there is merit in the "our" prefix here if that's the case.

tnull replied:
Good point, also the variable name should therefore maybe be called `receiver_node_id`?

tee8z replied:
maybe I should drop the `i.e., our node` part of the comment as well as change the variable name to receiver_node_id?

tnull replied:
Yeah, seems reasonable.

tnull commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -1380,9 +1380,10 @@ macro_rules! expect_payment_received {
 		assert_eq!(events.len(), 1);
 		match events[0] {
-			$crate::util::events::Event::PaymentReceived { ref payment_hash, ref purpose, amount_msat } => {
+			$crate::util::events::Event::PaymentReceived {ref payment_hash, ref purpose, amount_msat, our_node_id } => {
```
nit:
```suggestion
			$crate::util::events::Event::PaymentReceived { ref payment_hash, ref purpose, amount_msat, our_node_id } => {
```

tnull commented on lightning/src/util/events.rs:
```diff
@@ -273,6 +275,8 @@ pub enum Event {
 	///
 	/// [`ChannelManager::claim_funds`]: crate::ln::channelmanager::ChannelManager::claim_funds
 	PaymentClaimed {
+		/// The node that recieved the payment, ie our node
```
nit:
```suggestion
		/// The ID of the node that received the payment, i.e., our node.
```

tnull commented on lightning/src/util/events.rs:
```diff
@@ -631,6 +635,7 @@ impl Writeable for Event {
 					(4, amount_msat, required),
 					(6, 0u64, required), // user_payment_id required for compatibility with 0.0.103 and earlier
 					(8, payment_preimage, option),
+					(10, our_node_id, option),
```
As Matt mentioned, these should be odd-typed fields, i.e., 11 here, 7 below.

jurvis commented on lightning/src/util/events.rs:
```diff
@@ -721,12 +726,13 @@ impl Writeable for Event {
 					(0, payment_hash, required),
 					(2, purpose, required),
 					(4, amount_msat, required),
+					(6, our_node_id, option),
```
I believe this should be 7

tee8z commented:
renamed `our_node_id` to `receiver_node_id` and updated the comments to reflect that change, along with addressed the nit comments.

jurvis commented on lightning/src/util/events.rs:
```diff
@@ -612,7 +616,7 @@ impl Writeable for Event {
 				// drop any channels which have not yet exchanged funding_signed.
 			},
-			&Event::PaymentReceived { ref payment_hash, ref amount_msat, ref purpose } => {
+			&Event::PaymentReceived { ref payment_hash, ref amount_msat, ref purpose, receiver_node_id } => {
```
nit: receiver_node_id doesn't need an explicit `ref` here, but may be nice for consistency.

jurvis reviewed (dismissed):
Non-blocking nit for me

tnull commented on lightning/src/util/events.rs:
```diff
@@ -249,6 +249,8 @@ pub enum Event {
 	/// [`ChannelManager::claim_funds`]: crate::ln::channelmanager::ChannelManager::claim_funds
 	/// [`ChannelManager::fail_htlc_backwards`]: crate::ln::channelmanager::ChannelManager::fail_htlc_backwards
 	PaymentReceived {
+		/// The node that received the payment
```
nit:
```suggestion
		/// The node that received the payment.
```

tnull commented on lightning/src/util/events.rs:
```diff
@@ -273,6 +275,8 @@ pub enum Event {
 	///
 	/// [`ChannelManager::claim_funds`]: crate::ln::channelmanager::ChannelManager::claim_funds
 	PaymentClaimed {
+		/// The node that received the payment
```
nit:
```suggestion
		/// The node that received the payment.
```

tnull reviewed (commented):
Two tiny nits, LGTM otherwise.

tee8z commented:
Hey @TheBlueMatt Yes! Sorry for the delay, been diving into other things and forgot I still had this open, I'll get to it over the next few days

tee8z commented:
oh boy, messed that up, hold on

tee8z commented:
So I'm having issues trying to find the best way to support the test  `test_phantom_failure_reject_payment()` in `onion_route_tests.rs` as it's using the `expect_payment_received!` macro which always expects the event to have the node_id's real pubkey and not the phantom's pubkey. I'm wondering if the best choice would be to make a new macro that expects a phantom_pubkey (may need to add a function to node to expose this some how), or add a toggle to the existing one. Any direction here would be really helpful. Otherwise, I think I addresses all your comments @TheBlueMatt but please let me know anything I missed, thank you!

tee8z commented:
very cool! okay will do

tnull commented on lightning/src/util/events.rs:
```diff
@@ -1117,13 +1134,17 @@ impl MaybeReadable for Event {
+						(7, receiver_node_id, option),
 					});
 					if purpose.is_none() { return Ok(None); }
+
```
nit: this line can be removed.
```suggestion
```

tnull commented on lightning/src/util/events.rs:
```diff
@@ -1117,13 +1134,17 @@ impl MaybeReadable for Event {
 						(0, payment_hash, required),
 						(2, purpose, ignorable),
 						(4, amount_msat, required),
+						(7, receiver_node_id, option),
```
This also probably should be the next odd type, i.e., 1?

tnull commented on lightning/src/util/events.rs:
```diff
@@ -923,13 +937,15 @@ impl MaybeReadable for Event {
 						(4, amount_msat, required),
 						(6, _user_payment_id, option),
 						(8, payment_preimage, option),
+						(11, receiver_node_id, option),
```
This probably should be the next odd type, i.e., 1? (Although, depends if this will be merged before or after #1856, as this also adds some new fields here.

tnull commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3484,8 +3484,15 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
+											let phantom_shared_secret = claimable_htlc.prev_hop.phantom_shared_secret;
 											htlcs.push(claimable_htlc);
+											let mut receiver_node_id = self.our_network_pubkey;
+											if phantom_shared_secret.is_some(){
```
nit: add space
```suggestion
											if phantom_shared_secret.is_some() {
```

tnull commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3526,8 +3533,16 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
+														let phantom_shared_secret = &claimable_htlc.prev_hop.phantom_shared_secret.clone();
 														e.insert((purpose.clone(), vec![claimable_htlc]));
+														let mut receiver_node_id = Some(self.our_network_pubkey);
+														if phantom_shared_secret.is_some(){
```
nit: add space
```suggestion
														if phantom_shared_secret.is_some() {
```

tnull commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3526,8 +3533,16 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
+														let mut receiver_node_id = Some(self.our_network_pubkey);
+														if phantom_shared_secret.is_some(){
+															let phantom_pubkey = self.keys_manager.get_node_id(Recipient::PhantomNode)
+																				.expect("Failed to get node_id for phantom node recipient");
```
nit: remove space.
```suggestion
																.expect("Failed to get node_id for phantom node recipient");
```

tnull commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4225,6 +4241,12 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
+				let phantom_shared_secret = htlc.prev_hop.phantom_shared_secret;
+				if phantom_shared_secret.is_some(){
+					let phantom_pubkey = self.keys_manager.get_node_id(Recipient::PhantomNode)
+										.expect("Failed to get node_id for phantom node recipient");
```
nit: remove space
```suggestion
						.expect("Failed to get node_id for phantom node recipient");
```

tnull commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7487,6 +7510,13 @@ impl<'a, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
+					let phantom_shared_secret = claimable_htlcs[0].prev_hop.phantom_shared_secret;
+					if phantom_shared_secret.is_some(){
+						let phantom_pubkey = args.keys_manager.get_node_id(Recipient::PhantomNode)
+										.expect("Failed to get node_id for phantom node recipient");
```
nit: remove space.
```suggestion
						.expect("Failed to get node_id for phantom node recipient");
```

tnull commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7487,6 +7510,13 @@ impl<'a, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 					let mut claimable_amt_msat = 0;
+					let mut receiver_node_id = Some(our_network_pubkey);
+					let phantom_shared_secret = claimable_htlcs[0].prev_hop.phantom_shared_secret;
+					if phantom_shared_secret.is_some(){
```
nit: add space.
```suggestion
					if phantom_shared_secret.is_some() {
```

tnull commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -1433,20 +1433,22 @@ macro_rules! expect_pending_htlcs_forwardable_from_events {
 	}}
 }
-
+// expect_payment_received!(&nodes[fwd_idx], payment_hash, payment_secret, payment_amt, payment_preimage_opt);
```
This line should probably be removed?
```suggestion
```

tnull commented on lightning/src/util/events.rs:
```diff
@@ -758,6 +770,7 @@ impl Writeable for Event {
 					(4, amount_msat, required),
 					(6, 0u64, required), // user_payment_id required for compatibility with 0.0.103 and earlier
 					(8, payment_preimage, option),
+					(11, receiver_node_id, option),
```
See other comments: should likely be 1.

tnull commented on lightning/src/util/events.rs:
```diff
@@ -854,12 +867,13 @@ impl Writeable for Event {
 					(0, payment_hash, required),
 					(2, purpose, required),
 					(4, amount_msat, required),
+					(7, receiver_node_id, option),
```
See other comments: should likely be 1.

tnull reviewed (commented):
Did a quick pass, mostly some formatting nits.

tnull reviewed (commented):
Generally looks good and pretty close to being good to go.

One suggestion and a some formatting nits from my side, feel free to squash any additional changes you make right away so we can land this soon!

tnull commented on lightning/src/util/events.rs:
```diff
@@ -342,6 +342,13 @@ pub enum Event {
 	PaymentReceived {
+		/// The node that received the payment.
+		/// This is useful to identify payments which were received via [phantom node payments].
+		/// This field will always be filled in when the event was generated by LDK 
```
Super-tiny nit: pushing `versions` to the next line here and below looks a bit off (and introduced trailing whitespaces again). Feel free to leave as is though, if no other changes are requested.

tee8z replied:
hmm, so maybe I just keep it as one long line? not sure where else in that sentence to break on to keep it under the desired character limit

tnull replied:
Huh? If you'd keep "versions" on the previous line, my editor shows a line width of 92?

tee8z replied:
I was trying to follow this guideline: https://github.com/lightningdevkit/rust-lightning/pull/1766#pullrequestreview-1186534248 but will update to move `versions` to the line below, no biggie

tnull reviewed (approved):
LGTM, one tiny, non-blocking nit.

tee8z commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3445,7 +3445,13 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
+									receiver_node_id = self.keys_manager.get_node_id(Recipient::PhantomNode)
+										.expect("Failed to get node_id for phantom node recipient");
+								}
+								
```
so I shouldn't re-add the blank line before the macro then?
